phoenix-core-server/src/main/java/org/apache/phoenix/replication/ReplicationLogManager.java

```diff
@@ -0,0 +1,579 @@
+        "phoenix.replication.log.rotation.time.ms";
+    public static final long DEFAULT_REPLICATION_LOG_ROTATION_TIME_MS = 60 * 1000L;
+    /** Replication log rotation size trigger, default is 256 MB */
+    public static final String REPLICATION_LOG_ROTATION_SIZE_BYTES_KEY =
+        "phoenix.replication.log.rotation.size.bytes";
+    public static final long DEFAULT_REPLICATION_LOG_ROTATION_SIZE_BYTES = 256 * 1024 * 1024L;
+    /** Replication log rotation size trigger percentage, default is 0.95 */
+    public static final String REPLICATION_LOG_ROTATION_SIZE_PERCENTAGE_KEY =
+        "phoenix.replication.log.rotation.size.percentage";
+    public static final double DEFAULT_REPLICATION_LOG_ROTATION_SIZE_PERCENTAGE = 0.95;
+    /** Replication log compression, default is "NONE" */
+    public static final String REPLICATION_LOG_COMPRESSION_ALGORITHM_KEY =
+        "phoenix.replication.log.compression";
+    public static final String DEFAULT_REPLICATION_LOG_COMPRESSION_ALGORITHM = "NONE";
+
+    public static final String SHARD_DIR_FORMAT = "shard-%04d";
+    public static final String FILE_NAME_FORMAT = "%d-%s.plog";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReplicationLogManager.class);
+
+    private static volatile ReplicationLogManager instance;
+
+    private final Configuration conf;
+    private final RegionServerServices rsServices;
+    protected FileSystem standbyFs;
+    protected FileSystem fallbackFs; // For store-and-forward (future use)
+    protected int numShards;
+    protected URI standbyUrl;
+    protected URI fallbackUrl; // For store-and-forward (future use)
+    private final long rotationTimeMs;
+    private final long rotationSizeBytes;
+    private final Compression.Algorithm compression;
+    private final ReentrantLock lock = new ReentrantLock();
+    protected volatile LogFile.Writer currentWriter; // Current writer
+    //private volatile LogFile.Writer activeWriter; // Current active side writer (future use)
+    //private volatile LogFile.Writer fallbackWriter; // Current fallback writer (future use)
+    protected final AtomicLong lastRotationTime = new AtomicLong();
+    protected final AtomicLong writerGeneration = new AtomicLong();
+    private ScheduledExecutorService rotationExecutor;
+    private final AtomicBoolean started = new AtomicBoolean(false);
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final ConcurrentHashMap<Path,Object> shardMap = new ConcurrentHashMap<>();
+
+    /**
+     * Gets the singleton instance of the ReplicationLogManager using the lazy initializer pattern.
+     * Initializes the instance if it hasn't been created yet.
+     * @param conf Configuration object.
+     * @return The singleton ReplicationLogManager instance.
+     * @throws IOException If initialization fails.
+     */
+    public static ReplicationLogManager getInstance(Configuration conf,
+            RegionServerServices rsServices) throws IOException {
```

Comment:
The IndexRegionObserver coproc has the RegionCoprocessorEnvironment interface. I didn't find any way to get RegionServerServices from that interface. https://github.com/apache/phoenix/blob/master/phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java#L407

Reply:
Ok, see 2bd6d010a . Instead of aborting the regionserver if there is a Disruptor level problem we force close the writer. Instead of having the writer do a server abort the responsibility is pushed up to the higher layers, the users of the writer, which will receive IOExceptions because the writer has been closed. This provides more flexibility. (We can try to roll the writer first?) Although the challenge of aborting the regionserver if that must happen falls to the callers.